### PR TITLE
Add reusable block tab to inserter.

### LINKED ISSUE #22860: Find a better way to present reusable blocks in the inserter
Currently, reusable blocks are really buried in the inserter. They're located at the bottom of the "Blocks" tab, and with no way to filter or collapse categories, you have to scroll really far to even see them.

I think reusable blocks ought to be split out into their own primary category in the inserter, similar to block patterns. It's worth noting that template parts will likely get their own primary category soon.

The problem, of course, is that there's almost no room for any more tabs, especially when you take into account some languages using much longer words for certain phrases.

Because of this, I think we need to move away from tabs to something with more room for expansion. My current thinking is that a `<select>` dropdown would work well; it would make it easy to fit in 4 or more general categories, e.g.:

- Blocks
- Patterns
- Reusable Blocks
- Template Parts

I consider this a high priority issue, since due to the lack of collapsible block categories in the current inserter, it is way harder to find reusable blocks in `master` compared to current core WP.

### LINKED ISSUE #23253: Editor: Move Reusable blocks section up within Block Inserter menu
Had mentioned this in my other Issue #23076 (More menu improvement suggestions), but posting again here so it can be addressed separately with some more detail:

I think we should move the Reusable blocks section up in hierarchy within the Block Inserter, since users will be creating these on their own and may more likely reuse them compared to other block types in the Inserter. This section should also still include the "Manage all reusable block" text link at the bottom of it.

Screengrab GIFs of the [top bar block inserter](https://d.pr/i/gZHSur) and [inline block inserter](https://d.pr/i/RiAAAP) for context.

**Before/After:**
Screen Shot 2020-06-17 at 12 18 42 PM

### PR DESCRIPTION
**Pull request by ZebulanStanphill**
## Description

closes #22860.
closes #23253

This PR moves reusable blocks from a category mostly hidden at the bottom of the standard blocks tab in the inserter to their own tab.

Currently, it uses the style of the block tab for presenting the reusable blocks. But I think it may make more sense to present them the same way as patterns. Let me know what you think in #22860.

## How has this been tested?
I tested to make sure the tab showed reusable blocks, that inserting reusable blocks worked, and that the reusable category was removed from the standard blocks tab.

## Screenshots 
image

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by MichaelArestad** [comment]
This works well. My only suggestion from a design perspective is to change the label, as @mtias suggested, to "Saved."

**Comment by ZebulanStanphill** [comment]
Does that make sense? It seems weird and inconsistent to call them "Saved" in one spot but "Reusable" everywhere else.

**Comment by michelleweber** [comment]
I tend to lean toward sticking with "Reusable" for consistency -- the existence/idea of reusable blocks is still new, so I'd want people to be confident in their ability to find what they need by looking for the "reusable" language that we use elsewhere.

**Comment by mtias** [comment]
I think we should rename to _Saved_ across the board if we lift it up as its own grouping (which I like). I've seen _Reusable_ to suffer from lack of context (reusable blocks is much better than reusable) and translation shortcomings in a few languages. Saved stands on its own a bit better next to Patterns and Blocks.

That said, we should look at renaming separately and review the different contexts in which we show the action, as well as the icon.

File: packages/block-editor/src/components/inserter/menu.js
Comment (mtias) [inline packages/block-editor/src/components/inserter/menu.js:150]: Comment refers to patterns

File: packages/block-editor/src/components/inserter/reusable-block-list.js
Comment (mtias) [inline packages/block-editor/src/components/inserter/reusable-block-list.js:70]: Is this container necessary?

File: packages/block-editor/src/components/inserter/reusable-block-list.js
Comment (mtias) [inline packages/block-editor/src/components/inserter/reusable-block-list.js:54]: Not for this PR, but I think we should let go of "reusable" being a category and instead allow the tab to display other categories. That would allow a site to categorize reusable blocks (with the regular category interface when managing the custom post type) and have the categories listed in this tab as regular block categories.

Comment (ZebulanStanphill) [reply packages/block-editor/src/components/inserter/reusable-block-list.js:54]: Agreed.

File: packages/block-editor/src/components/inserter/reusable-block-list.js
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/inserter/reusable-block-list.js:70]: I'm not sure. I was just copying what `block-list.js` did.

File: packages/block-editor/src/components/inserter/menu.js
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/inserter/menu.js:150]: Thanks for catching.

**Comment by enriquesanchez** [comment]
Tested this PR on Safari + VoiceOver. The tabs are properly announced as before, both their name and the number of tab (ie. 3 of 3), as well as the number of blocks under each one (with the exception of Patterns).

From an accessibility perspective, this looks good to me 👍

**Comment by ZebulanStanphill** [comment]
After #22789 got merged, I needed to rebase this. I did, and now the reusable blocks tab isn't showing up. I'll let you know when I fix it.

**Comment by ZebulanStanphill** [comment]
Alright, everything should be working again.

File: packages/block-editor/src/components/inserter/menu.js
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/inserter/menu.js:58]: Should I instead use the following?

```js
select( 'core' ).getEntityRecords( 'postType', 'wp_block' )?.length
```

I'm not sure which makes more sense or which is faster.

File: packages/block-editor/src/components/inserter/reusable-block-list.js
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/inserter/reusable-block-list.js]: I've made some adjustments to clean up the DOM, so the extra `<div>`s are gone now.

**Comment by ZebulanStanphill** [comment]
Alright, I've spent the whole day working on fixing the end-to-end tests. I'm down to just one failing test, and I can't figure out why it's failing:

```
FAIL packages/e2e-tests/specs/editor/plugins/cpt-locking.test.js (43.064s)
  ● cpt locking › template_lock all › can use the global inserter in inner blocks

    expect(received).not.toBeNull()

    Received: null

      120 | 			await pressKeyTimes( 'Tab', 2 );
      121 | 			await page.keyboard.press( 'Enter' );
    > 122 | 			expect( await page.$( '.wp-block-gallery' ) ).not.toBeNull();
          | 			                                                  ^
      123 | 		} );
      124 | 	} );
      125 | 

      at Object.<anonymous> (specs/editor/plugins/cpt-locking.test.js:122:54)
          at runMicrotasks (<anonymous>)
```

It seems like the Reusable block tab is being added (causing the inserter to use tabs, and thus causing an extra <kbd>Tab</kbd> press to be necessary), but if that's the case (and it might not be), I don't know what's causing the Reusable block tab to appear. I need help to figure this one out.

**Comment by jasmussen** [comment]
I took a look at this, and tests pass for me locally. 

I even tried this interactive test:

```
npm run test-e2e packages/e2e-tests/specs/editor/plugins/cpt-locking.test.js  -- --puppeteer-interactive
```

and the failing test in question appears to work as intended visually too. Here's the test bit:

```
			await page.click( '.edit-post-header-toolbar__inserter-toggle' );
			await page.type(
				'.block-editor-inserter__search-input',
				'gallery'
			);
			await pressKeyTimes( 'Tab', 2 );
			await page.keyboard.press( 'Enter' );
			expect( await page.$( '.wp-block-gallery' ) ).not.toBeNull();
```

So, click the inserter, search for "gallery", press tab twice, and you should've inserted a gallery block. This is where it gets interesting, because regardless of configuration, this always takes 3 tab presses for me. Here's your branch with no reusable blocks:
  
no reusables

Here's your branch, with a reusable block:

gallery

Here's the master branch with a reusable block:

master with reusable

In all three, you press tab 3 times. Which begs the questions:

- Why are the tests passing on master?
- Why are the tests passing for me on this branch, locally, but failing here?

@youknowriad — Jorge wrote these tests but is out. Do you have any insights here?

File: packages/block-editor/src/components/inserter/menu.js
Comment (youknowriad) [reply packages/block-editor/src/components/inserter/menu.js:58]: no, you used the right one here. The block-editor package should use the REST API direclty.

File: packages/block-editor/src/components/inserter/reusable-blocks-tab.js
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/reusable-blocks-tab.js:129]: Minor: It seems like having InsertableReusableBlocks and ReusableBlocksSearchResults components is not really necessary since both have a similar UI (single panel with all the blocks)

File: packages/block-editor/src/components/inserter/test/reusable-blocks-tab.js
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/test/reusable-blocks-tab.js:158]: I'm not really certain this unit tests is very valuable personally but your call.

File: packages/e2e-tests/specs/editor/various/reusable-blocks.test.js
Comment (youknowriad) [inline packages/e2e-tests/specs/editor/various/reusable-blocks.test.js:40]: I don't think these showBlockToolbar calls are necessary. `clickBlockToolbarButton` already  calls it.

**Comment by youknowriad** [comment]
I think the test failure comes from the fact that if you have reusable blocks, you'll have an extra tab stop on that inserter. 
Ideally, the reusable blocks are removed (like the post) at the beginning of end2end tests and end2end tests creating them should clean them when they finish.

**Review by youknowriad — COMMENTED** [review]
Cool stuff, this is looking good, I left some minor comments  and suggestions about the e2e tests errors.

File: packages/block-editor/src/components/inserter/test/reusable-blocks-tab.js
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/inserter/test/reusable-blocks-tab.js:158]: I didn't create these tests from scratch. They were originally in the unit test file for the standard blocks tab. I just moved them over since reusable blocks are now in their own tab.

File: packages/e2e-tests/specs/editor/various/reusable-blocks.test.js
Comment (ZebulanStanphill) [reply packages/e2e-tests/specs/editor/various/reusable-blocks.test.js:40]: Thanks for catching that!

**Comment by youknowriad** [comment]
Small tip: instead of amending commits (sometimes it's necessary though), it's better to keep pushing new commits in PRs. This makes it easier for folks following the PR to know what changed in the latest commits.

File: packages/block-editor/src/components/inserter/reusable-blocks-tab.js
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/reusable-blocks-tab.js:41]: Minor: if you filter the reusable block items before calling `searchBlockItems`, you'd be able to remove the duplication and inline `reusableBlockFilter`

**Comment by ZebulanStanphill** [comment]
I didn't realize reusable blocks were persisted across tests. Is there a simple way to delete all reusable blocks at the end of a test? Or do you have to perform the deletion "manually" (by opening the more menu on each reusable block and clicking the "Remove from Reusable blocks" option).

**Comment by youknowriad** [comment]
@ZebulanStanphill There's a `trashExistingPosts` function in `setup-test-framework.js`, we should do something similar for reusable blocks. (Potentially we could make it a utility)

**Review by youknowriad — APPROVED** [review]
Let's get this in to be ready for 5.5 beta 1.

💯 kudos here @ZebulanStanphill

**Comment by jasmussen** [comment]
👌

**Comment by ZebulanStanphill** [comment]
Thanks for fixing the tests, @youknowriad! :smile:

**Pull request merged**
